### ISSUE TITLE
Ruby process hangs when calling generator from after_bundle hook in Rails template

### ISSUE THREAD

**Issue by jensljungblad** [rails/rails#21700]
If I do the following in a Rails template:
```ruby
after_bundle do
  generate(:model, "article title:string")
end
```

And run if with
```shell
$ rails new sandbox -m template.rb
```

The ruby process hangs afterwords. For the record, I'm not trying to generate a model, but I'm trying to run a generator that comes from a gem that I'm adding in the template as well.

**Comment by zerothabhishek** [comment]
Tried what you mentioned, and the two processes were indeed running as you say. But could not get the crash on running a similar rails new command -

    rails new sandbox -m template.rb   #=> sandbox app generated, two spring processes running
    rails new sandbox2 -m template.rb  #=> sandbox2 app generated, two more spring processes running

Generators generally start spring while loading the app, and don't attempt to shut it down after done. That's kind of expected, can we call it a bug?

**Comment by jensljungblad** [comment]
You're right, that does work. Seems the problem comes when I delete a generated project and run the generator again (which you normally wouldn't do except when testing your template I'm guessing)

So the following crashes, at least on OS X:
```
rails new sandbox -m template.rb
rm -fr sandbox
rails new sandbox -m template.rb
```

**Comment by maclover7** [comment]
This still seems to be happening in Rails 5, as reported via #27178.

**Comment by VishalRocks** [comment]
Is this a synchronization problem as two processes running together is the problem? Should we use a mutex?

**Comment by CharlieHawker** [comment]
For those looking for a workaround, simply adding `run "spring stop"` at the top of your `after_bundle` block seems to work nicely. Also fixes another issue I had where an initial git commit inside the block didn't pick up the binstubs.

I will try to work on a PR if I get a chance but that seems to have done the trick for me.

Here's my block now:

```
...

after_bundle do
  run "spring stop" # Fix pesky hangtime

  # Install rspec, simple_form & devise
  generate("rspec:install")
  generate("simple_form:install --foundation")
  generate("devise:install")

  # Run erb2haml
  rails_command "haml:erb2haml"

  # Create the default action / view
  route "root to: 'home#index'"
  generate("controller", "home index")

  # Initialise git
  git :init
  git add: "."
  git commit: %Q{ -m 'Initial commit' }
end

...
```

**Comment by danielricecodes** [comment]
@CharlieHawker's workaround looks pretty good to me.  The issue appears to be rooted in using a preloader, `spring`, and a template at the same time.  This issue could, or should, probably be closed.

**Comment by viktorianer** [comment]
Any progress on this issue? 

I found it, because I am running in a similar issue now. I would like to use `after_bundle` in a Rails template, e.g. this simple one on an existing application:

`rails app:template LOCATION='https://railsbytes.com/public/templates/zNPs7l'`

If I add a `after_bundle` block to this template it is not executed at all. No matter what this block does. I think that should be fixed?!

**Comment by gerrywastaken** [comment]
@viktorianer That's a different bug which I've recently submitted a PR to address: https://github.com/rails/rails/pull/40197
